### DISCUSSION TITLE
Adding new metrics after registering collector

### DISCUSSION THREAD

**Discussion by piotrpio** [prometheus/client_golang#1421] [Q&A]
I am writing a custom exporter with collectors discovering and adding new metrics based on response from the server. Up until now, we were simply sending a single request when the exporter starts up and based on the response we were adding metrics and then registering collectors. However, we have a use case where not all metrics we may want to track are available when the exporter first starts, so I would like to be able to add new metrics to the collector during scrape (basically in `Collect()`).

Simply creating e.g. `NewGaugeVec()`  with config for the new metric seems to work, but there is a problem when I try to unregister the collector - since I added new metrics and those are exposed in my `Describe()`  implementation, the `collectorID`  calculated in `func (r *Registry) Unregister(c Collector) bool` will be different from the one calculated when first registering the collector. Because of that, the collector is never removed from collectors map in `Registry` (https://github.com/prometheus/client_golang/blob/main/prometheus/registry.go#L383).
Is my approach to the problem even valid? If so, do you have any suggestions how this problem can be solved?

**Comment by piotrpio** [comment]
Any updates on this? If it's a bug (not recalculating the ID), I can issue a fix, but I did not want to attempt that unless it's confirmed...

**Comment by bwplotka** [comment]
Hello from the Bug Scrub!

Is this still relevant? Sorry for lag.

I think there are two ways you can accomplish this now AFAIK:
* Just do the [unchecked collector](https://github.com/prometheus/client_golang/blob/a4af5c85ff987d05e390802916f4d8b5bc7b61a2/prometheus/collector.go#L38), so don't provide any definitions on `Describe`, but instead collect static (const) metrics
* You could perhaps maintain a dynamic list of things in `Describe` and the moment this list changes, manually register those and append to the list. This way, when unregistering collector, things are cleaned up, no?

Hope that helps!
